### PR TITLE
Update QDQ propagation transformer to insert QDQ nodes

### PR DESCRIPTION
**Pull request by edgchen1**
**Description**
Update QDQ propagation transformer to insert new QDQ nodes instead of moving the existing one. This creates a more consistent `DQ -> op -> Q` pattern for other components to recognize.
Upgrade this transformer to a basic level optimization as it yields a valid ONNX graph.

**Motivation and Context**
Enable the effects of the QDQ propagation transformer in minimal builds by making it a basic level optimization. It can run prior to saving the ORT format model.

### CODE REVIEW

File: onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc [hunk 1]
```diff
@@ -1,139 +1,277 @@
...
+  // there is also the case where the graph input is an output, but we don't care about that for this transformer
+};
+
+bool CanNodePropagate(const Node& node) {
   return graph_utils::IsSupportedOptypeVersionAndDomain(node, "MaxPool", {12}) ||
          graph_utils::IsSupportedOptypeVersionAndDomain(node, "Reshape", {5, 13, 14}) ||
          graph_utils::IsSupportedOptypeVersionAndDomain(node, "Transpose", {1, 13});
 }
 
-static bool TryCancelOutDQQPair(Graph& graph, Node& dq_node, Node& q_node) {
```
Comment (edgchen1) [inline onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc:21]: @yufenglee This change removes the DQ/Q cancellation. This transformer shouldn't create any such pairs, but matching DQ/Q pairs in the original model will no longer be removed. Is this fine?

File: onnxruntime/core/optimizer/qdq_transformer/qdq_s8_to_u8.cc [hunk 2]
```diff
@@ -1,18 +1,20 @@
...
 
 namespace onnxruntime {
 
 // Convert QuantizeLinear and DequantizeLinear pair with type int8_t to type uint8_t
-Status QDQS8ToU8Transformer::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+Status QDQS8ToU8Transformer::ApplyImpl(Graph& graph, bool& modified, int graph_level,
+                                       const logging::Logger& logger) const {
   GraphViewer graph_viewer(graph);
-  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+  const auto node_topology_list = gsl::make_span(graph_viewer.GetNodesInTopologicalOrder());
```
Comment (edgchen1) [inline onnxruntime/core/optimizer/qdq_transformer/qdq_s8_to_u8.cc:17]: >make_span [](http://example.com/codeflow?start=39&length=9)

revert this

File: onnxruntime/core/optimizer/qdq_transformer/qdq_util_minimal.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+// 2. scale and zero point is constant scalar
+// 3. Q and DQ have same scale and zero point
+bool IsQDQPairSupported(
+    const Node& q_node, const Node& dq_node,
+    const std::function<const ONNX_NAMESPACE::TensorProto*(const std::string&)>& get_const_initializer,
+    const Path& model_path);
+
+// Check if DQ is supported in the QDQ transformer. It requires:
+// 1. DQ doesn't have optional input.
+// 2. scale and zero point is constant scalar
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/qdq_transformer/qdq_util_minimal.h:41]: not understanding this limitation. the zero point value is the only optional input, and if missing defaults to 0. why do we care if it's there or not?

File: onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc [hunk 4]
```diff
@@ -1,139 +1,277 @@
...
 #include "core/graph/graph_utils.h"
 #include "core/optimizer/initializer.h"
 #include "core/optimizer/qdq_transformer/qdq_util.h"
 #include "core/optimizer/utils.h"
 
 namespace onnxruntime {
+namespace {
+// like graph_utils::GraphEdge, but also allowing the source end to be an input and the destination end to be an output
+struct ExtendedGraphEdge {
+  enum class End { Source,
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc:17]: Should this go in graph_utils? Seems fairly general purpose and other optimizers may be able to benefit from it in the future.

File: onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc [hunk 5]
```diff
@@ -1,139 +1,277 @@
...
+
+  bool HasGraphInputOrInitializer() const noexcept { return !src.has_value(); }
+  bool HasGraphOutput() const noexcept { return !dst.has_value(); }
+
+  const std::optional<NodeInfo>& GetNodeInfoAtEnd(End end) const {
+    return end == End::Source ? src : dst;
+  }
+
+  Node* GetNodeAtEnd(Graph& graph, End end) const {
+    if (const auto& node_info = GetNodeInfoAtEnd(end); node_info.has_value()) {
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc:37]: Should this method be non-const? If not do we need two different GetNodeAtEnd methods?

File: onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc [hunk 6]
```diff
@@ -157,41 +295,44 @@ bool QDQPropagationTransformer::PropagateDQForward(Graph& graph) const {
       continue;
     }
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc:296]: Could IsDQSupported be used here?

File: onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc [hunk 7]
```diff
@@ -211,49 +352,49 @@ bool QDQPropagationTransformer::PropagateQBackward(Graph& graph) const {
...
     ORT_RETURN_IF_ERROR(Recurse(*node_ptr, modified, graph_level, logger));
   }
 
-  while (PropagateQBackward(graph) || PropagateDQForward(graph)) {
-  }
+  const auto& compatible_eps = GetCompatibleExecutionProviders();
 
-  modified = true;
+  ORT_RETURN_IF_ERROR(PropagateQBackward(graph, node_indices, compatible_eps, logger, modified));
+  ORT_RETURN_IF_ERROR(PropagateDQForward(graph, node_indices, compatible_eps, logger, modified));
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc:397]: If we're inserting pairs do we need a forward and backwards pass?

File: onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc (hunk 6 above)
Comment (edgchen1) [reply onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc:296]: it looks like it. maybe I can add a similar IsQSupported function too

File: onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc (hunk 7 above)
Comment (edgchen1) [reply onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc:397]: DQ forward pass handles: QDQ -> (op(s) that can propagate) -> (op that cannot propagate)
Q backward pass handles: (op that cannot propagate) -> (op(s) that can propagate) -> QDQ
so I think both directions are needed, unless I'm missing something

File: onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc (hunk 4 above)
Comment (edgchen1) [reply onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc]: moved it to a separate file and put it in the graph_utils namespace

File: onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc (hunk 5 above)
Comment (edgchen1) [reply onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc]: renamed it to GetMutableNodeAtEnd. I think it's ok as a const method as it doesn't actually expose modifiable state of the ExtendedGraphEdge, but rather the non-const Graph parameter

File: onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc (hunk 6 above)
Comment (edgchen1) [reply onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc]: renamed it and used it in both propagate functions

**Review by skottmckay — APPROVED** [review]
:shipit:

File: onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc (hunk 1 above)
Comment (yufenglee) [reply onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc:21]: yes, it is good as we won't insert QDQ if the tensor is already quantized.

**Review by yufenglee — DISMISSED** [review]
:shipit:

File: onnxruntime/core/optimizer/qdq_transformer/qdq_util_minimal.h (hunk 3 above)
Comment (yufenglee) [reply onnxruntime/core/optimizer/qdq_transformer/qdq_util_minimal.h]: code in selector assumes no optional input, so we add the constraint here. And we will insert a zero point manually if it is optional.

Comment (edgchen1) [reply onnxruntime/core/optimizer/qdq_transformer/qdq_util_minimal.h]: when all the QDQ transformers are at level 2 this makes sense. but now with the QDQ propagation one moving to level 1 I'll update it to handle missing zero point input.

**Review by yufenglee — APPROVED** [review]
:shipit:

**Pull request merged**